### PR TITLE
Send telemetry event from SharedTree Rebase operation.

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

[ADO 8083](https://dev.azure.com/fluidframework/internal/_workitems/edit/8083/)

#### Related Items

[ADO 5141](https://dev.azure.com/fluidframework/internal/_workitems/edit/5141/)
[ADO 8152](https://dev.azure.com/fluidframework/internal/_workitems/edit/8152/)
https://github.com/microsoft/FluidFramework/pull/21325

This PR:
- Implements logger in SharedTree's `rebase()` method to send telemetry event.
- Extracts `measure()` method outside to `utils`.

### CODE REVIEW

File: packages/dds/tree/src/shared-tree-core/editManager.ts [hunk 1]
```diff
@@ -133,23 +139,38 @@ export class EditManager<
...
 		this.sequenceMap.set(minimumPossibleSequenceId, this.trunkBase);
+
+		if (this.logger !== undefined) {
+			this.telemetryEventBatcher = new TelemetryEventBatcher(
+				{
+					eventName: "rebaseProcessing",
+					category: "performance",
+				},
+				this.logger,
+				1000, // TODO: Replace pattern.
```
Comment (alexvy86) [inline packages/dds/tree/src/shared-tree-core/editManager.ts:157]: Not sure what "replace pattern" means, but do we have a number we want to go with for the batching right now? @yann-achard-MS you might be the one who know best if a hardcoded number here would be serviceable to start. How many rebase operations do we expect a given client to see normally, such that we get some messages out of this, but not too many? In other places we have something [like this](https://github.com/alexvy86/FluidFramework/blob/1c6c72689e27b8c140d6250814481d26ae097132/packages/dds/shared-object-base/src/sharedObject.ts#L162) that gives us flexibility to tweak the number (getting a mc/monitoringContext just requires a call [like this](https://github.com/alexvy86/FluidFramework/blob/1c6c72689e27b8c140d6250814481d26ae097132/packages/dds/shared-object-base/src/sharedObject.ts#L133) around the logger).

File: packages/utils/telemetry-utils/src/utils.ts [hunk 2]
```diff
@@ -104,3 +104,14 @@ export function createSampledLogger(
 
 	return sampledLogger;
 }
+
+/**
+ * TODO
+ * @public
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/utils.ts:110]: Internal for sure.
```suggestion
 * Runs the specified function and returns an object with the time it took to run as well as any output from it.
 *
 * @remarks Useful in conjunction with {@link TelemetryEventBatcher}.
 * @internal
```

File: packages/utils/telemetry-utils/src/telemetryEventBatcher.ts [hunk 3]
```diff
@@ -13,32 +11,16 @@ import type {
...
-export interface IMeasuredCodeResult<TKey extends string> {
-	/**
-	 * Optional properties to log custom data. The set of properties must be the same for all calls to the `measure` function.
-	 */
-	telemetryProperties?: { readonly [key in TKey]: number };
-}
-
-/**
- * Telemetry class that measures the execution time of a given piece of code and accumulates user defined telemetry metrics, to finally log an event through the {@link TelemetryEventBatcher.logger | logger} provided to this class when the number of calls to the {@link TelemetryEventBatcher.measure | measure} function reaches the specified by {@link TelemetryEventBatcher.threshold | threshold}.
+ * Telemetry class that measures the execution time of a given piece of code and accumulates user defined telemetry metrics, to finally log an event through the {@link TelemetryEventBatcher.logger | logger} provided to this class when the number of calls to the function reaches the specified by {@link TelemetryEventBatcher.threshold | threshold}.
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/telemetryEventBatcher.ts:14]:
```suggestion
 * Telemetry class that accumulates measurements which are eventually logged in a telemetry event through the provided
 * {@link TelemetryEventBatcher.logger | logger} when the number of calls to the function reaches the specified {@link TelemetryEventBatcher.threshold | threshold}.
```

File: packages/utils/telemetry-utils/src/telemetryEventBatcher.ts [hunk 4]
```diff
@@ -71,34 +53,13 @@ export class TelemetryEventBatcher<TMetrics extends string> {
...
-	}
-
 	/**
 	 * Accumulates the custom data and sends it to the logger every {@link TelemetryEventBatcher.threshold} calls.
 	 *
 	 * @param customData -
 	 * A record storing the custom data to be accumulated and eventually logged.
 	 */
-	private accumulateAndLog(customData: Record<TMetrics, number>): void {
+	public accumulateAndLog(customData: Record<TMetrics, number>): void {
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/telemetryEventBatcher.ts:62]: One small quirk of the event logged by this class is that it won't have a `duration` property, but `avgDuration` and `maxDuration`. It's unconventional but fine since `duration` is optional in the performance event interface, but maybe something we want to come back to at some point.

File: packages/utils/telemetry-utils/src/telemetryEventBatcher.ts
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/telemetryEventBatcher.ts:1]: I think Josh brought this up in the original PR, in lines 85 and 91 we probably should do `avg_${key}` and `max_${key}`, otherwise since we're not capitalizing the first letter of the keys we'll have `maxduration` and `avgduration` instead of `maxDuration` and `avgDuration`, etc.

File: packages/dds/tree/src/shared-tree-core/editManager.ts (hunk 1 above)
Comment (jikim-msft) [reply packages/dds/tree/src/shared-tree-core/editManager.ts:157]: The first link you attached is exactly what I meant by 😄. Working on it right now.

File: packages/utils/telemetry-utils/src/telemetryEventBatcher.ts
Comment (jikim-msft) [reply packages/utils/telemetry-utils/src/telemetryEventBatcher.ts:1]: This has also been included in my latest commit 😄. Telepathy!

File: packages/test/test-service-load/src/utils.ts [hunk 5]
```diff
@@ -302,6 +302,7 @@ export async function safeExit(code: number, url: string, runId?: number) {
 export const globalConfigurations: Record<string, ConfigTypes> = {
 	"Fluid.SharedObject.DdsCallbacksTelemetrySampling": 10000,
 	"Fluid.SharedObject.OpProcessingTelemetrySampling": 10000,
+	"Fluid.SharedTree.RebaseSampling": 10000,
```
Comment (jikim-msft) [inline packages/test/test-service-load/src/utils.ts:305]: @yann-achard-MS 

Do you have a number in mind that you would like to have?

Comment (alexvy86) [reply packages/test/test-service-load/src/utils.ts:305]: @jikim-msft this is the one I'm saying we should remove, this is specifically for stress tests. The 1000 in editManager.ts is the one we should just make sure is at a value we like.

Comment (jikim-msft) [reply packages/test/test-service-load/src/utils.ts:305]: Makes sense. Thanks for catching this. Would it be okay to create a constant variable inside the `editManager.ts` (e.g., `const THRESHOLD = 1000`) and pass that to the logger?

In this case, we can remove the monitoringContext.

Comment (alexvy86) [reply packages/test/test-service-load/src/utils.ts:305]: If we're very confident with the default it would be fine, but I'd say there's no harm in still using the configurable value from monitoringContext and just not overriding it anywhere, letting it be undefined there and default to 1000 or whatever we end up coalescing to in that line.

Comment (yann-achard-MS) [reply packages/test/test-service-load/src/utils.ts:305]: I think that for production 1000 would be good (wherever that value is supposed to be)

File: packages/dds/tree/src/shared-tree-core/editManager.ts [hunk 6]
```diff
@@ -133,23 +140,40 @@ export class EditManager<
 		public readonly changeFamily: TChangeFamily,
 		public readonly localSessionId: SessionId,
 		private readonly mintRevisionTag: () => RevisionTag,
+		private readonly logger?: ITelemetryLoggerExt,
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/editManager.ts:143]: It looks like we don't need to keep the logger instance as a member of this class.

File: packages/dds/tree/src/shared-tree-core/editManager.ts [hunk 7]
```diff
@@ -133,23 +140,40 @@ export class EditManager<
...
 	) {
 		this.trunkBase = {
 			revision: "root",
 			change: changeFamily.rebaser.compose([]),
 		};
 		this.sequenceMap.set(minimumPossibleSequenceId, this.trunkBase);
+
+		if (this.logger !== undefined) {
+			const THRESHOLD = 1000;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/editManager.ts:152]: If we need to keep a constant in this file, let's make it a module-scope one instead of redeclaring a variable here.

File: packages/dds/tree/src/shared-tree-core/editManager.ts (hunk 6 above)
Comment (alexvy86) [reply packages/dds/tree/src/shared-tree-core/editManager.ts:143]:
```suggestion
		readonly logger?: ITelemetryLoggerExt,
```

**Review by alexvy86 — APPROVED** [review]
Looks good from my end, let's wait for that last look from Yann.

File: experimental/dds/tree/src/SharedTree.ts [hunk 8]
```diff
@@ -827,10 +827,7 @@ export class SharedTree extends SharedObject<ISharedTreeEvents> implements NodeI
 	public loadSummary(summary: SharedTreeSummaryBase): void {
 		const { version: loadedSummaryVersion } = summary;
 
-		if (
-			this.deltaManager.readOnlyInfo.readonly !== true &&
-			isUpdateRequired(loadedSummaryVersion, this.writeFormat)
-		) {
+		if (this.deltaManager.readOnlyInfo.readonly !== true && isUpdateRequired(loadedSummaryVersion, this.writeFormat)) {
```
Comment (yann-achard-MS) [inline experimental/dds/tree/src/SharedTree.ts:830]: It would be best to revert this change too.

File: packages/dds/tree/src/shared-tree-core/editManager.ts [hunk 9]
```diff
@@ -40,12 +41,22 @@ import {
...
 
 export const minimumPossibleSequenceNumber: SeqNumber = brand(Number.MIN_SAFE_INTEGER);
 const minimumPossibleSequenceId: SequenceId = {
 	sequenceNumber: minimumPossibleSequenceNumber,
 };
 
+/**
+ * Amount of time the telemetry logger will wait before sending a batch of events.
+ */
+const THRESHOLD = 1000;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/editManager.ts:58]:
```suggestion
/**
 * Max number of telemetry log call that may be aggregated before being sent.
 */
const maxRebaseStatsAggregationCount = 1000;
```

File: packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts [hunk 10]
```diff
@@ -141,6 +145,11 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
 			policy: schemaPolicy,
 		};
 
+		this.rebaseLogger = createChildLogger({
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:148]: This is not being used. You need to pass it to the EditManager constructor.

Comment (yann-achard-MS) [reply packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:148]: Also I don't think this class really needs to keep this as a member.

File: packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts [hunk 11]
```diff
@@ -69,13 +73,13 @@ export interface ClonableSchemaAndPolicy extends SchemaAndPolicy {
...
 	private readonly editManager: EditManager<TEditor, TChange, ChangeFamily<TEditor, TChange>>;
 	private readonly summarizables: readonly Summarizable[];
-
 	/**
 	 * The sequence number that this instance is at.
 	 * This number is artificial in that it is made up by this instance as opposed to being provided by the runtime.
 	 * Is `undefined` after (and only after) this instance is attached.
 	 */
 	private detachedRevision: SeqNumber | undefined = minimumPossibleSequenceNumber;
+	rebaseLogger: ITelemetryLoggerExt;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:82]: We don't need this

Comment (jikim-msft) [reply packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:82]: I'm currently building my code with format and necessary changes. Just one second!

**Review by yann-achard-MS — APPROVED** [review]
